[RFC] provider: Only call system() once, don't use Python 3 interpreter for +python, improve messages

fwalch opened the pull request:
The executable 'python' can refer to either Python 2 or Python 3. Add a
check to only accept Python 2 interpreters as providers for +python.

Also improve the error messages.

Resolves #2734.

_Edit:_ now includes #2549.

blueyed committed:
provider/pythonx: only call system("python") once

`s:check_interpreter` will query/return the version also, so that
`s:check_version` can just use that, without calling "python" again.

blueyed committed:
provider/pythonx: fix grammar with error

blueyed committed:
provider/pythonx: add reference to ":help nvim-python" for error

blueyed committed:
provider/pythonx: test python3/2 first, do not test python3.2

python3/python2 will typically point at the default / most recent
interpreter.

blueyed committed:
provider/pythonx: Detect: only return valid versions

The `check` return value of `s:check_version` was not being used, and it
was returned always.

blueyed committed:
provider/python: define Prog and Error functions always

`provider#pythonX#Error` and `provider#pythonX#Prog` are currently meant
for debugging only (the error message is not being used), and should
therefore be defined always, especially in case of errors.

Ref: https://github.com/neovim/neovim/pull/2549#issuecomment-98607580

blueyed committed:
provider/python: Call: fix usage of 'finish' in function

Do not call it again in case of an exception in `remote#host#Require`
(ref: https://github.com/neovim/neovim/pull/2549#issuecomment-102674350).

blueyed committed:
remote#host#RequirePythonHost: fix reference to log

I could not find a reference to NVIM_PYTHON_PYTHON_LOG anywhere, and
python-client looks for NVIM_PYTHON_LOG_FILE.

~/.nvimlog appears to be hardcoded and enabled by default.  This would
need to be adjusted when this changes.

blueyed committed:
provider/pythonx: return error from Detect

Ref: https://github.com/blueyed/neovim/commit/9b653ce0#commitcomment-11195449

fwalch committed:
provider/pythonx: Don't use Python 3 interpreter for +python.

The executable 'python' can refer to either Python 2 or Python 3. Add a
check to only accept Python 2 interpreters as providers for +python.

Also improve the error messages.

Resolves #2734.

fwalch commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -13,25 +13,36 @@ function! provider#pythonx#Detect(ver) abort
-        \   ['2.7', '2.6', '2', '']
-        \ : ['3.5', '3.4', '3.3', '3.2', '3', '']
+        \   ['2', '']
+        \ : ['3', '']
```
I think we should only check the default Pythons to not increase confusion.

Shougo replied:
I think your fix should be merged in https://github.com/neovim/neovim/pull/2549.
And please consider about https://github.com/neovim/neovim/pull/1334#issuecomment-94178512 problem.

fwalch commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -13,25 +13,36 @@ function! provider#pythonx#Detect(ver) abort
-  return ['', 'Neovim module installed Python'
-        \ .a:ver.' interpreter is not found.']
+  " No suitable Python interpreter found.
+  return ['', full_err]
```
I'm not sure how to even look at the output of this (i.e. the contents of `full_err`), but the output of `echom` looked good.

fwalch commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -43,15 +54,17 @@ function! s:check_version(prog, ver, skip) abort
   let min_version = (a:ver == 2) ? '2.6' : '3.3'
-  if system(a:prog . get_version) >= min_version
+  let prog_version = system(a:prog . get_version)
+  if prog_version =~ '^' . a:ver && prog_version >= min_version
```
This is the real change in this PR: check that the major version of Python is the requested one. Before, Python 3 was accepted as a Python 2 interpreter (since 3.x >= 2).

Shougo replied:
Oh, I get it.  It is problem.

fmoralesc commented:
LGTM.

justinmk commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -43,15 +54,17 @@ function! s:check_version(prog, ver, skip) abort
 function! s:check_interpreter(prog, ver, skip) abort
   if !executable(a:prog)
-    return [0, 'Python'.a:ver.' interpreter is not executable.']
+    return [0, 'The Python interpreter ' . a:prog . ' is not executable.']
```
I suggest leaving out "The", it's almost always a waste of space.

fmoralesc replied:
In that case, wouldn't it be better to phrase it: `'provider/pythonx: '. a.prog . ' is not executable.'`?

Shougo replied:
@fmoralesc I agree with the prefix.

justinmk replied:
:+1:

justinmk commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -64,6 +77,7 @@ function! s:check_interpreter(prog, ver, skip) abort
         \   '''import importlib; exit(importlib.find_loader("neovim") is None)''')
         \ )
-  return [!v:shell_error, 'Python'.a:ver.' interpreter have not neovim module.']
+  return [!v:shell_error, 'The Python interpreter ' . a:prog .
```
Don't need "The".

justinmk commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -43,15 +54,17 @@ function! s:check_version(prog, ver, skip) abort
   endif
-  return [0, 'Python ' . get_version . ' interpreter is not supported.']
+  return [0, 'The Python ' . prog_version . ' interpreter ' . a:prog .
+        \ ' is not suitable for providing support for Python ' . a:ver . '.']
```
Wording is arguably more confusing now. Suggestion:

    Python X cannot provide Python Y.

fmoralesc replied:
Likewise, `provider/python: '.a:prog.' cannot provide Python '.a:ver.'.'`

justinmk replied:
:+1:

blueyed commented:
As @Shougo pointed out, this need to get rebased on #2549, once it's merged.

fwalch commented:
Rebased on top of #2549. Changed the messages according to the comments. As all error messages are collected, the prefix was only added [here](https://github.com/neovim/neovim/commit/3e0c0d596c41863affbda72cac354d3c9c739696#diff-88fff3c190a530ba2b28d59291d12a33R48).

fwalch commented:
Are the error messages returned by the Python detection (`provider#python#Error()`) actually used/stored/displayed somewhere? I can't see them in `:messages`.

blueyed commented:
> Are the error messages returned by the Python detection (provider#python#Error()) actually used/stored/displayed (apart from s:err in the provider code) somewhere? I can't see them in :messages.

No.  See https://github.com/neovim/neovim/pull/2549#issuecomment-98418657.

fwalch committed:
provider/pythonx: Merge s:check_version() and s:check_interpreter().

fwalch commented:
Added a commit to merge the `s:check_version()` and `s:check_interpreter()` functions, as IMO it doesn't make sense to have them as separate functions (they were called in immediate succession).

fwalch commented:
Amended @blueyed's commit to use `echomsg` and `echohl WarningMsg` (ref https://github.com/neovim/neovim/pull/2549#issuecomment-108957283). Also changed one of my own commits to use lists and `join()` instead of string concatenation.

fwalch commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -5,65 +5,71 @@ endif
-  return ['', 'Neovim module installed Python'
-        \ .a:ver.' interpreter is not found.']
+  " No suitable Python interpreter found.
+  return ['', 'provider/pythonx: Could not load Python ' . a:major_ver
```
Sample output:

```
provider/pythonx: Could not load Python 2:                                                                                                                     
python2 does have not have the neovim module installed. See ":help nvim-python".                                                                               
python2.7 does have not have the neovim module installed. See ":help nvim-python".                                                                             
python2.6 does not exist or is not executable.                                                                                                                 
python is Python 3.4 and cannot provide Python 2.                                                                                                              
```

fwalch committed:
provider/pythonx: Show absolute path to executable in error message.

fwalch commented:
Added a new commit to show absolute paths in the error messages to make debugging PATH issues easier.

Shougo commented:
:+1:

fwalch commented:
I know this is getting a bit unwieldy, so to summarize for reviewers:
 * #2549 has been pulled in unchanged, except https://github.com/neovim/neovim/pull/2735#issuecomment-108973336.
 * Three commits added on top of that (probably easiest to review by commit). These could be sent as individual PRs, if you prefer that. I only pulled in #2549 to to already resolve merge conflicts, but then https://github.com/neovim/neovim/pull/2549#issuecomment-107810536 came up.

justinmk commented:
LGTM.

fwalch commented:
Rebased. Will merge after the build has finished.
